### ISSUE TITLE
Typedef and compatibility

### ISSUE THREAD

**Issue by Remi123** [odinthenerd/tmp#5]
Hi Odin, I hope you have a great day, I'm really impress by some of your work.

It's not really an issue, but since I'm stuck at work with GCC 4.3 -std=gnu++0x, I did some experimentation to adapt your code to this environment. To my surprise, I did manage to replace a lot of using statement to their typedef counterpart :
`template<typename T,typename C = identity_>
struct dispatch<1,is_<T,C>>
{ template<typename A>
using f = typename dispatch... f<std::is_same<T,A>>;};`
become
`template<typename T,typename C = identity_>
struct dispatch<1,is_<T,C>>
{ template<typename A>
struct f  {typedef typename dispatch... f<std::is_same<T,A>::type>::type type;
};};`

This add a lot of boilerplate but I was able to make it compatible with GCC 4.3 since it have variadic template. The only difference is that listify_ is now a specialization of dispatch implemented with lift_<list,C>, and I must provide typename C = listify_<> as a default template metaclosure, but it get the job done. ALso most constexpr function like find_dispatch() must be reimplemented as a uint_<(calculation ? ...)>::value.

I only remade simple construct until now and I'm seeing some progress, but I may not see the big picture on where I'll hit a wall due to some properties of using statement.
Fusion-based features were an important part of your talks, and I'm not even scratching the surface of those properties which I doubt could work at compile time under my conditions, but    
 it's worth a try. 

My issue is about the development of this : Do you foresee problem using this method and are you interested in seeing a merge request ? I also don't know if you are rewriting things up since the number of commits is low, as you would have yet again found another template metatrick.

Thanks and I really enjoy your conference,
Remi123

**Comment by Quincunx271** [comment]
I strongly suspect that this is not worth it. Odin's TMP libraries are based around avoiding as many template instantiations as possible via using alias templates, but if you replace the alias templates with structs like this, you lose that benefit.

In short, I suspect that you'll get dismal compile times, and you should switch to an easier-to-understand style of metaprogramming library, rather than this arcane tacit style. Remember the Rule of Chiel: https://www.reddit.com/r/cpp/comments/6gur2x/the_rule_of_chiel_aka_compiletime_cost_of/

**Comment by odinthenerd** [comment]
I think mp11 is probably your best bet for GCC4.3

**Comment by Remi123** [comment]
Hi again,
I was able to continue a little bit of conversion just to see if it still worked under much more pressure. After implementing make_sequence, I could do some test with O(N) algo like filter. The compile time is similar, although I must be careful not to hit the templated recursion limit. 
The much bigger problem I faced was the lack of constexpr. Each int_ is a static const int which work when calling ::value but I saw a huge memory footprint in the elf file each time the algo found an integral_contant. GCC 4.5 optimize them away but not GCC 4.3. I needed to specialize dispatch to accept my struct along with some template specialization over  some template template metafunction like add_
```
template<int V>
struct i_{ typedef i_ type;};

template<typename U, typename V>
struct add_;
template<int U, int V>
struct add_<i_<U>, i_<V>> 
    : i_<(U+V)>
{};

 ```
and reformulate my continuation to accept only bool_<true> or bool_<false> since I can afford two static bool. The memorization of the integral_constant does happen but they are instantiated anyway. So without this "optimization", I'll have mangled name integral_constant<int> from 0 to my max and two integral_constant<bool>.

Something like : 
```      
    // Let's redo the famous Gauss problem where you need to add each numbers between 
    // 1 and 100.
static_assert(
          call_<result_<is_<i_<5050>>>,
            call_<make_sequence_<identity_,fold_left_<lift_<add_>>>,
                i_<101> // make_sequence start at 0 and finish a N-1
        >>::type::value ,"");

    // But the Gauss solution was to multiply 101 by 50
    static_assert(
        call_<result_<is_<i_<5050>>>,
            call_<lift_<multiply_>,
                i_<101>, i_<50>
                >>::type::value  ,"");
```
This is not the best implementation but it work and I'm still learning the syntax.
@Quincunx271 It does work extremely fast for our needs, although we didn't benchmark it against the original yet. I'm suspect I got 50% more compile time at worst, but I'm not even sure that the compiler doesn't do that under the hood. The way I see it, I just create a huge nested typedef typename dispatch<...>::type type resolved by call_



Thanks again,

Remi123
